--- a/include/uapi/linux/netfilter/nfnetlink.h	2015-08-07 15:31:14.307624417 +0800
+++ b/include/uapi/linux/netfilter/nfnetlink.h	2015-08-07 15:31:42.989624219 +0800
@@ -51,6 +51,7 @@
 #define NFNL_SUBSYS_ACCT		7
 #define NFNL_SUBSYS_CTNETLINK_TIMEOUT	8
 #define NFNL_SUBSYS_CTHELPER		9
-#define NFNL_SUBSYS_COUNT		10
+#define NFNL_SUBSYS_URLSET		10
+#define NFNL_SUBSYS_COUNT		11
 
 #endif /* _UAPI_NFNETLINK_H */
