--- a/kernel/net/netfilter/ipset/ip_set_hash_ip.c	2016-08-29 19:42:29.694899476 +0800
+++ b/kernel/net/netfilter/ipset/ip_set_hash_ip.c	2016-08-29 19:46:23.772898936 +0800
@@ -17,6 +17,7 @@
 #include <net/ipv6.h>
 #include <net/netlink.h>
 #include <net/tcp.h>
+#include <linux/if_arp.h>
 
 #include <linux/netfilter.h>
 #include <linux/netfilter/ipset/pfxlen.h>
@@ -87,8 +88,22 @@
 	struct hash_ip4_elem e = { 0 };
 	struct ip_set_ext ext = IP_SET_INIT_KEXT(skb, opt, set);
 	__be32 ip;
+	struct arphdr *arp_h = NULL;
+	unsigned char *arp_ptr = NULL;	
+	
+	/* ETH_P_ARP only for the garp defense, 'ip' stores the sender ip.  --2015.05.18*/
+	if(htons(ETH_P_ARP) == skb->protocol)
+	{
+		arp_h = arp_hdr(skb);
+		arp_ptr = (unsigned char *)(arp_h + 1);
+		arp_ptr += ETH_ALEN;
+		memcpy(&ip, arp_ptr, 4);
+	}
+	else
+	{
+		ip4addrptr(skb, opt->flags & IPSET_DIM_ONE_SRC, &ip);
+	}
 
-	ip4addrptr(skb, opt->flags & IPSET_DIM_ONE_SRC, &ip);
 	ip &= ip_set_netmask(h->netmask);
 	if (ip == 0)
 		return -EINVAL;
