--- linux-3.3.8.orig/net/netfilter/nf_conntrack_netlink.c	2016-03-22 09:26:19.061831081 +0800
+++ linux-3.3.8/net/netfilter/nf_conntrack_netlink.c	2016-03-22 09:36:34.858821414 +0800
@@ -559,6 +559,11 @@
 void (*g_appdist_event)(unsigned int events, struct nf_ct_event *item);
 EXPORT_SYMBOL(g_appdist_event);
 
+void (*g_special_route_create)(unsigned int events, struct nf_ct_event *item);
+EXPORT_SYMBOL(g_special_route_create);
+void (*g_special_route_destroy)(unsigned int events, struct nf_ct_event *item);
+EXPORT_SYMBOL(g_special_route_destroy);
+
 #ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
 static int ctnetlink_conntrack_event(struct notifier_block *this,
                            unsigned long events, void *ptr)
@@ -591,10 +596,18 @@
 		{
 			(*g_appdist_event)(events, item);
 		}
+		if(g_special_route_destroy)
+		{
+			(*g_special_route_destroy)(events, item);
+		}
 	} else  if (events & ((1 << IPCT_NEW) | (1 << IPCT_RELATED))) {
 		type = IPCTNL_MSG_CT_NEW;
 		flags = NLM_F_CREATE|NLM_F_EXCL;
 		group = NFNLGRP_CONNTRACK_NEW;
+		if(g_special_route_create)
+		{
+			(*g_special_route_create)(events, item);
+		}
 	} else  if (events) {
 		type = IPCTNL_MSG_CT_NEW;
 		group = NFNLGRP_CONNTRACK_UPDATE;
