--- a/net/8021q/Kconfig	2012-06-01 15:16:13.000000000 +0800
+++ b/net/8021q/Kconfig	2017-05-19 14:39:23.855192507 +0800
@@ -27,3 +27,15 @@
 	  automatic propagation of registered VLANs to switches.
 
 	  If unsure, say N.
+
+ 
+config VLAN_MULTICAST_IPTV
+	bool "Multicast IPTV"
+	depends on VLAN_8021Q
+	---help---
+	  If you say Y here, then the IPTV will do with multicast and unicast vlan id
+	  based on VLAN information in the packet.
+         
+	  Say N to exclude this support and reduce the binary size.
+         
+	  If unsure, say Y.
--- a/net/8021q/vlan_core.c	2017-04-06 17:35:50.627197609 +0800
+++ b/net/8021q/vlan_core.c	2017-05-19 14:41:48.543194021 +0800
@@ -5,6 +5,11 @@
 #include <linux/export.h>
 #include "vlan.h"
 
+#ifdef CONFIG_VLAN_MULTICAST_IPTV 
+u16 (*iptv_handle_frame_hook)(struct sk_buff *skb, u16 vlan_tci, int flag) __read_mostly;
+EXPORT_SYMBOL_GPL(iptv_handle_frame_hook);
+#endif
+
 bool vlan_do_receive(struct sk_buff **skbp, bool last_handler)
 {
 	struct sk_buff *skb = *skbp;
@@ -132,6 +137,14 @@
 
 	vhdr = (struct vlan_hdr *) skb->data;
 	vlan_tci = ntohs(vhdr->h_vlan_TCI);
+	#ifdef CONFIG_VLAN_MULTICAST_IPTV
+	if (iptv_handle_frame_hook)
+	{
+		u16 tmp_tci;
+		tmp_tci  = iptv_handle_frame_hook(skb, vlan_tci, 0);
+		vlan_tci = tmp_tci?tmp_tci:vlan_tci;
+	} 
+	#endif 
 	__vlan_hwaccel_put_tag(skb, vlan_tci);
 
 	skb_pull_rcsum(skb, VLAN_HLEN);
