--- a/drivers/mtd/devices/m25p80.c	2015-10-15 11:59:50.000000000 +0800
+++ b/drivers/mtd/devices/m25p80.c	2016-03-30 17:20:15.000000000 +0800
@@ -749,6 +749,7 @@
 	/* GigaDevice */
 	{ "gd25q32", INFO(0xc84016, 0, 64 * 1024,  64, SECT_4K) },
 	{ "gd25q64", INFO(0xc84017, 0, 64 * 1024, 128, SECT_4K) },
+	{ "gd25q256c", INFO(0xc84019, 0, 64 * 1024, 512, SECT_4K) },
 
 	/* Intel/Numonyx -- xxxs33b */
 	{ "160s33b",  INFO(0x898911, 0, 64 * 1024,  32, 0) },
