Index: linux-3.10.14/drivers/net/raeth/raether.c
===================================================================
--- linux-3.10.14.orig/drivers/net/raeth/raether.c	2014-12-03 10:07:13.603710204 +0800
+++ linux-3.10.14/drivers/net/raeth/raether.c	2014-12-03 10:12:34.124249140 +0800
@@ -873,8 +873,12 @@
 #if defined (CONFIG_RAETH_SKB_RECYCLE_2K)
                 skb = skbmgr_dev_alloc_skb2k();
 #else
+#if LINUX_VERSION_CODE > KERNEL_VERSION(3,10,0)
+		skb = alloc_skb(MAX_RX_LENGTH + NET_IP_ALIGN, GFP_ATOMIC);
+#else
 		skb = __dev_alloc_skb(MAX_RX_LENGTH + NET_IP_ALIGN, GFP_ATOMIC);
 #endif
+#endif
 
 		if (unlikely(skb == NULL))
 		{
