--- a/arch/mips/kernel/setup.c	2019-01-21 16:10:25.418205652 +0800
+++ b/arch/mips/kernel/setup.c	2019-04-24 08:55:24.822204402 +0800
@@ -649,6 +649,13 @@
 	pr_info("@@----debug cmdline=[ %s ]\n", command_line);
 #endif
 
+
+#ifdef CONFIG_MTK_MACH_OSG_V1
+	strlcpy(command_line, " board=OSGv1 console=ttyS1,115200 root=31:02 init=/sbin/init mtdparts=raspi:256k(u-boot)ro,3072k(kernel),12288k(rootfs)", COMMAND_LINE_SIZE);
+
+	pr_info("@@----debug cmdline=[ %s ]\n", command_line);
+#endif
+
 	*cmdline_p = command_line;
 
 	parse_early_param();
--- a/arch/mips/ralink/Kconfig	2019-01-21 16:10:25.419205652 +0800
+++ b/arch/mips/ralink/Kconfig	2019-04-24 08:56:04.233201631 +0800
@@ -433,4 +433,5 @@
 config MTK_MACH_TL_R480_V9
 	bool "TP-LINK TL-R480 V9 support"
 
-
+config MTK_MACH_OSG_V1
+	bool "TP-LINK OSGv1 support"
--- a/arch/mips/ralink/Makefile	2019-01-21 16:10:25.419205652 +0800
+++ b/arch/mips/ralink/Makefile	2019-04-24 08:56:53.201202192 +0800
@@ -44,4 +44,6 @@
 obj-$(CONFIG_RALINK_MT7628) +=  ehci_ohci.o udc.o dev-gpio-buttons.o dev-leds-gpio.o
 obj-$(CONFIG_MTK_MACH_TL_R470_V6) += mach-tl-r470-v6.o
 obj-$(CONFIG_MTK_MACH_TL_R480_V9) += mach-tl-r480-v9.o
+obj-$(CONFIG_RALINK_MT7621) +=  ehci_ohci.o udc.o dev-gpio-buttons.o dev-leds-gpio.o
+obj-$(CONFIG_MTK_MACH_OSG_V1) += mach-osg-v1.o
 
