diff -Naur a/vconfig.c b/vconfig.c
--- a/networking/vconfig.c	2015-10-11 17:53:46.000000000 +0800
+++ b/networking/vconfig.c	2015-10-11 18:03:21.000000000 +0800
@@ -19,6 +19,7 @@
 //usage:     "\n	set_egress_map	VLAN_NAME SKB_PRIO VLAN_QOS"
 //usage:     "\n	set_ingress_map	VLAN_NAME SKB_PRIO VLAN_QOS"
 //usage:     "\n	set_name_type	NAME_TYPE"
+//usage:     "\n        set_untag       IFACE 0|1"
 
 #include "libbb.h"
 #include <net/if.h>
@@ -32,7 +33,11 @@
 	GET_VLAN_INGRESS_PRIORITY_CMD,
 	GET_VLAN_EGRESS_PRIORITY_CMD,
 	SET_VLAN_NAME_TYPE_CMD,
-	SET_VLAN_FLAG_CMD
+        SET_VLAN_FLAG_CMD,
+        GET_VLAN_REALDEV_NAME_CMD, /* If this works, you know it's a VLAN device, btw */
+        GET_VLAN_VID_CMD, /* Get the VID of this VLAN (specified by name) */
+        SET_VLAN_UNTAG,
+        GET_VLAN_UNTAG
 };
 enum vlan_name_types {
 	VLAN_NAME_TYPE_PLUS_VID, /* Name will look like:  vlan0005 */
@@ -53,6 +58,7 @@
 		unsigned int name_type;
 		unsigned int bind_type;
 		unsigned int flag; /* Matches vlan_dev_info flags */
+                unsigned int untag;
 	} u;
 
 	short vlan_qos;
@@ -112,10 +118,21 @@
 	++argv;
 	if (!argv[0])
 		bb_show_usage();
+
+        if (!strcmp(argv[0],"set_untag"))
+        {
+                ifr.cmd = SET_VLAN_UNTAG;
+                if (argc != 4)
+                        bb_show_usage();
+        }
+        else 
+        {
 	p = xfind_str(cmds + 2, argv[0]);
 	ifr.cmd = *p;
 	if (argc != p[-1])
 		bb_show_usage();
+        }
+        
 
 	if (ifr.cmd == SET_VLAN_NAME_TYPE_CMD) {
 		/* set_name_type */
@@ -130,11 +147,15 @@
 		 * doing so wouldn't save that much space and would also make maintainence
 		 * more of a pain.
 		 */
+                
 		if (ifr.cmd == SET_VLAN_FLAG_CMD) {
 			/* set_flag */
 			ifr.u.flag = xatou_range(p, 0, 1);
 			/* DM: in order to set reorder header, qos must be set */
 			ifr.vlan_qos = xatou_range(argv[3], 0, 7);
+                } else if (ifr.cmd == SET_VLAN_UNTAG) {
+                        /* set_untag */
+                        ifr.u.untag = xatou_range(p, 0, 1);
 		} else if (ifr.cmd == ADD_VLAN_CMD) {
 			/* add */
 			ifr.u.VID = xatou_range(p, 0, VLAN_GROUP_ARRAY_LEN - 1);
