--- a/networking/route.c	2017-05-10 19:18:50.184193738 +0800
+++ b/networking/route.c	2017-05-10 19:20:45.551147752 +0800
@@ -189,7 +189,10 @@
 			int prefix_len;
 
 			prefix_len = xatoul_range(prefix+1, 0, 32);
-			mask_in_addr(*rt) = htonl( ~(0xffffffffUL >> prefix_len));
+			if (prefix_len >= 32)
+				mask_in_addr(*rt) = htonl( 0xffffffffUL );
+			else
+				mask_in_addr(*rt) = htonl( ~(0xffffffffUL >> prefix_len));
 			*prefix = '\0';
 #if HAVE_NEW_ADDRT
 			rt->rt_genmask.sa_family = AF_INET;
