--- a/arch/arm/boot/dts/qcom-ipq8064-ap145.dtsi	2017-07-16 17:02:54.962737204 +0800
+++ b/arch/arm/boot/dts/qcom-ipq8064-ap145.dtsi	2017-07-16 17:05:09.518737738 +0800
@@ -475,6 +475,15 @@
                         phy4: ethernet-phy@4 {
                                 device_type = "ethernet-phy";
                                 reg = <4>;
+                                qca,ar8327-initvals = <
+                                        0x00004 0x7600000   /* PAD0_MODE */
+                                        0x00008 0x1000000   /* PAD5_MODE */
+                                        0x0000c 0x80        /* PAD6_MODE */
+                                        0x000e4 0xaa545     /* MAC_POWER_SEL */
+                                        0x000e0 0xc74164de  /* SGMII_CTRL */
+                                        0x0007c 0x4e        /* PORT0_STATUS */
+                                        0x00094 0x4e        /* PORT6_STATUS */
+                                >;
                         };
                 };
 
