--- a/arch/mips/ath79/dev-eth.c
+++ b/arch/mips/ath79/dev-eth.c
@@ -366,9 +366,11 @@
 {
 }
 
+#if 0
 static void ath79_ddr_no_flush(void)
 {
 }
+#endif
 
 static void ath79_ddr_flush_ge0(void)
 {
@@ -1004,13 +1006,15 @@
 			pdata->reset_bit = AR934X_RESET_GE0_MAC |
 					   AR934X_RESET_GE0_MDIO;
 			pdata->set_speed = ar934x_set_speed_ge0;
+			pdata->ddr_flush = ath79_ddr_flush_ge0;
 		} else {
 			pdata->reset_bit = AR934X_RESET_GE1_MAC |
 					   AR934X_RESET_GE1_MDIO;
 			pdata->set_speed = ath79_set_speed_dummy;
+			pdata->ddr_flush = ath79_ddr_flush_ge1;
 		}
 
-		pdata->ddr_flush = ath79_ddr_no_flush;
+		//pdata->ddr_flush = ath79_ddr_no_flush;
 		pdata->has_gbit = 1;
 		pdata->is_ar724x = 1;
		
--- a/drivers/net/ethernet/atheros/ag71xx/ag71xx.h
+++ b/drivers/net/ethernet/atheros/ag71xx/ag71xx.h
@@ -72,7 +72,7 @@
 #define AG71XX_HACK_WIFI_HEADROOM	128
 
 #define AG71XX_TX_RING_SIZE_DEFAULT	128
-#define AG71XX_RX_RING_SIZE_DEFAULT	128
+#define AG71XX_RX_RING_SIZE_DEFAULT	256
 
 #define AG71XX_TX_RING_SIZE_MAX		256
 #define AG71XX_RX_RING_SIZE_MAX		256
--- a/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c
+++ b/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c
@@ -129,12 +129,15 @@ static int ag71xx_ring_alloc(struct ag71
 	int i;
 
 	ring->desc_size = sizeof(struct ag71xx_desc);
+
+#if 0
 	if (ring->desc_size % cache_line_size()) {
 		DBG("ag71xx: ring %p, desc size %u rounded to %u\n",
 			ring, ring->desc_size,
 			roundup(ring->desc_size, cache_line_size()));
 		ring->desc_size = roundup(ring->desc_size, cache_line_size());
 	}
+#endif
 
 #ifdef CONFIG_AG71XX_SRAM_DESCRIPTORS
 	if (id < MAX_AG71XX_USING_SRAM) {
@@ -794,6 +797,7 @@ static netdev_tx_t ag71xx_hard_start_xmi
 
 	/* setup descriptor fields */
 	desc->data = (u32)dma_addr;
+	ag71xx_get_pdata(ag)->ddr_flush();
 	desc->ctrl = len & DESC_PKTLEN_M;
 
 	curr = curr->next;
@@ -1083,6 +1087,7 @@ static int ag71xx_rx_packets(struct ag71
 						rx_buf_size, DMA_FROM_DEVICE);
 
 		desc->data = (u32)curr->dma_addr;
+		wmb();
 		desc->ctrl = DESC_EMPTY;
 
 		/*
@@ -1168,7 +1173,9 @@ static int ag71xx_poll(struct napi_struc
 	int tx_done;
 	int rx_done;
 
+#if 0
 	pdata->ddr_flush();
+#endif
 
 	/*
 	 * First empty any packets that we have transmitted!  In theory it might
