--- a/arch/mips/include/asm/thread_info.h	2016-01-19 10:07:44.000000000 +0800
+++ b/arch/mips/include/asm/thread_info.h	2016-01-19 09:25:40.000000000 +0800
@@ -62,7 +62,7 @@
 
 /* thread information allocation */
 #if defined(CONFIG_PAGE_SIZE_4KB) && defined(CONFIG_32BIT)
-#define THREAD_SIZE_ORDER (1)
+#define THREAD_SIZE_ORDER (2)
 #endif
 #if defined(CONFIG_PAGE_SIZE_4KB) && defined(CONFIG_64BIT)
 #define THREAD_SIZE_ORDER (2)
