diff -purN a/file.c b/file.c
--- a/file.c	2017-08-09 10:18:30.444458942 +0800
+++ b/file.c	2017-08-09 10:18:53.976459898 +0800
@@ -684,8 +684,14 @@ static void uci_parse_option(struct uci_
 	else
 		UCI_NESTED(uci_set, ctx, &ptr);
 
-	if (!list && value[0])
-		ptr.o->encrypt_state = encrypt_state;
+	if (!list && value[0]) {
+		if (ctx->flags & UCI_FLAG_UPGRADE) {
+			if (ptr.o->encrypt_state == UCI_ENCRYPT_AUTO)
+				ptr.o->encrypt_state = encrypt_state;
+		} else {
+			ptr.o->encrypt_state = encrypt_state;
+		}
+	}
 }
 
 /*
