--- a/file.c	2018-10-17 09:27:51.650800533 +0800
+++ b/file.c	2018-10-17 10:10:10.930903781 +0800
@@ -1139,7 +1139,7 @@ static char **uci_list_config_files(stru
 {
 	char **configs;
 	glob_t globbuf;
-	int size, i;
+	int size, i, j;
 	char *buf;
 	char *dir;
 
@@ -1163,6 +1163,7 @@ static char **uci_list_config_files(stru
 
 	configs = uci_malloc(ctx, size);
 	buf = (char *) &configs[globbuf.gl_pathc + 1];
+	j = 0;
 	for(i = 0; i < globbuf.gl_pathc; i++) {
 		char *p;
 
@@ -1173,7 +1174,7 @@ static char **uci_list_config_files(stru
 		if (!uci_validate_package(p))
 			continue;
 
-		configs[i] = buf;
+		configs[j++] = buf;
 		strcpy(buf, p);
 		buf += strlen(buf) + 1;
 	}
