--- a/interface-ip.c	2016-01-11 20:12:12.000000000 +0800
+++ b/interface-ip.c	2016-04-12 15:34:04.241005810 +0800
@@ -571,8 +571,13 @@
 
 		if (!keep || replace) {
 			if (!(a_new->flags & DEVADDR_EXTERNAL)) {
-				if (system_add_address(dev, a_new))
-					a_new->failed = true;
+				if (system_add_address(dev, a_new)) {
+					/* try to delete duplicate address and add again */
+					system_del_address(dev, a_new);
+					if (system_add_address(dev, a_new)) {
+						a_new->failed = true;
+					}
+				}
 
 				if (iface->metric || a_new->policy_table)
 					interface_handle_subnet_route(iface, a_new, true);
