diff -uNr a/lua.c b/lua.c
--- a/lua.c	2017-05-10 09:38:04.510182569 +0800
+++ b/lua.c	2017-05-10 09:37:26.761182614 +0800
@@ -400,12 +400,15 @@
 {
 	char realm[256] = {};
 	char realm_decode[256] = {};
-	char ipc_auth_password[128] = {};
+	char ipc_auth_password[128] = {0};
 	char *ptr = NULL;
 	int len = 0;
 	int usr_len = 0;
 	struct blob_attr *cur;
 	int rem;
+	int i;
+	struct MD5Context md5_ctx;
+	unsigned char digest[16] = {0};
 
 	blob_for_each_attr(cur, cl->hdr.head, rem) {
 		if (!strncasecmp(blobmsg_name(cur), "authorization", 13)) {
@@ -415,7 +418,7 @@
 	}
 
 	usr_len = strlen(dev_usr);
-	if (len > 6 && len > usr_len) {
+	if (len > usr_len) {
 		char *tmp = NULL;
 
 		/* Pass the 'Basic ' */
@@ -431,7 +434,18 @@
 			if (!strcmp(ipc_auth_password, dev_password)) {
 				return 0;
 			}
-			else if (strcmp(dev_usr, "admin")){
+			memset(ipc_auth_password, 0, sizeof(ipc_auth_password));
+			MD5Init(&md5_ctx);
+			MD5Update(&md5_ctx, ptr, strlen(ptr));
+			MD5Final(digest, &md5_ctx);
+			for (i = 0; i < 16; i++) {
+				sprintf(ipc_auth_password + 2 * i, "%02x", *(digest + i));
+			}
+			if (!strcmp(ipc_auth_password, dev_password)) {
+				return 0;
+			}
+			
+			if (strcmp(dev_usr, "admin")){
 				char req[512],*resp = NULL;
 				cJSON * pJsonRoot = NULL,*pJsonErr =NULL;
 				int ret = -2,error_code = -1;
